draft-gpew-priv-ppm.md

```diff
@@ -628,268 +653,642 @@ encoded value of the following form:
+message as described below. Otherwise, if the leader finished or failed, then
+move to INVALID and proceed as described in {{aggregate-request}}.
 
-Once the aggregators have verified at least as many reports as required for the
-PPM task, the leader issues an "aggregate-share request" to each helper. The
-helper responds to this request by extracting its aggregate share from its state
-and encrypting it under the collector's HPKE public key.
+The helper computes its next state transition (denoted `prep_next`) as
 
-For each aggregator endpoint `[aggregator]` in the parameters associated with
-`CollectReq.task_id` (see {{pa-collect}}) except its own, the leader sends a
-POST request to `[aggregator]/aggregate_share` with the following message:
+~~~
+out = prep_state.next(payload)
+~~~
+
+where `payload` is the previous VDAF message sent by the leader. If this
+operation fails, then the helper fails with error `vdaf-prep-next-error`.
+Otherwise, it interprets `out` as follows. If this is the last round of VDAF
+preparation phase, then `out` is the helper's output share, in which case the
+helper finishes. Otherwise, the helper continues with `out` as its next VDAF
+message.
+
+### Aggregating Output Shares {#out-to-agg-share}
+
+Once processing of a report share reaches the FINISHED state, the aggreagtor
+stores the the recoverd output share until the batch to which it pertains is
+collected. To aggregate the output shares, the aggregator runs the aggreagtion
+algorithm specified by the VDAF:
+
+~~~
+agg_share = output_to_aggregate_shares(agg_param,
+                                       output_shares)
+~~~
+
+Note that for most VDAFs, it is possible to aggregate output shares as they
+arrive rather than wait until the batch is collected. To do so however, it is
+necessary to enforce the batch parameters as described in
+{{batch-parameter-validation}} so that the aggregator knows which aggregate
+share to update.
+
+The leader needs to store a report for as long as it can be collected. Once a
+report has been collected `max_batch_lifetime` times, the leader SHOULD remove
+the report from storage.
+
+### Aggregate Messages {#aggregate-message-auth}
+
+The following messages are defined in this section and are used to aggregate a
+nubmber of reports simultaneously:
+
+~~~
+enum {
+  agg_init_req(0),
+  agg_req(1),
+  agg_resp(2),
+  agg_share_req(3),
+  agg_share_resp(4),
+} AggregateType;
+
+struct {
+  AggregateType msg_type;
+  select (msg_type) {
+    agg_init_req:   AggregateInitReq;
+    agg_req:        AggregateReq;
+    agg_resp:       AggregateResp;
+    agg_share_req:  AggregateShareReq;
+    agg_share_resp: HpkeCiphertext;
+  }
+  opaque tag [32];
+} Aggregate;
+~~~
+
+The last 32 bytes of each message is an HMAC-SHA256 tag computed over the
+serialized message excluding the `tag` field itself. The key that is used is the
+`agg_auth_key` shared by the aggregators and configured for the given task.
```

Comment:
> Given that these messages will be transmitted over TLS, why do we need an additional MAC? My best guess is that it's to prove to the recipient that the message came from the other aggregator, but since the aggregators are already aware of each other's HPKE configs, could we use HPKE to [mutually authenticate](https://www.ietf.org/archive/id/draft-irtf-cfrg-hpke-12.html#name-authentication-using-an-asy)?

Some form of message authentication is needed because we don't assume mutually authenticated TLS. Yes, we could use one of the mutual auth modes for HPKE, but at the end of the day we would just use this to derive a MAC key.

> More broadly it feels wrong to put a transport level concern like integrity or authenticity in the message.

When we started working on this stuff, this was my inclination as well. However, conversations with Chris W., EKR, and others have lead me to think that there's value in minimizing the assumptions we make about what the underlying transport does, since it's evolution isn't tied to the evolution of our protocol.

> I think there's an analogy here to the problem of client<->aggregator authentication, in that I think the details of how a leader and helper authenticate to each other will be deployment specific and maybe we shouldn't try to specify it in PPM, but instead just abstractly declare that helper and leader must communicate over a confidential and mutually authenticated channel. mTLS works, but they could also use a PSK, or use a common identity provider, or who knows what.

I definitely see the value in keeping PPM agnostic about this. Many solutions are possible and not everyone is going to agree that a MAC is the best one. To be clear, I stuck a MAC here because the VDAF already requires distributing a shared secret key, so adding a MAC is cheap and convenient.

Reply:
We're going to do the MAC with a pre shared secret for now, and will revisit this as #161 and other issues related to sharing entropy between aggregators mature.